src/sentry/api/endpoints/custom_rules.py

```diff
@@ -261,41 +267,110 @@ def _rule_to_response(rule: CustomDynamicSamplingRule) -> Response:
+            and token.key.name == "event.type"
+            and token.value.value == "transaction"
+        ):
             transaction_filter = True
             break
+
     if not transaction_filter:
         return UnsupportedSearchQueryReason.NOT_TRANSACTION_QUERY
+
     return None
 
 
-def get_condition(query: Optional[str]) -> RuleCondition:
+def _message_to_transaction_condition(tokens: Sequence[QueryToken]) -> Sequence[QueryToken]:
+    """
+    Transforms queries containing messages into proper transaction queries
+
+    eg: "foo environment:development" -> "transaction:foo environment:development"
+
+    a string "foo" is parsed into a SearchFilter(key=SearchKey(name="message"), operator="=", value="foo")
+    we need to transform it into a SearchFilter(key=SearchKey(name="transaction"), operator="=", value="foo")
+
+    """
+    new_tokens: List[QueryToken] = []
+    for token in tokens:
+        if (
+            isinstance(token, SearchFilter)
+            and token.key.name == "message"
+            and token.operator == "="
+        ):
+            # transform the token from message to transaction
+            new_tokens.append(
+                SearchFilter(
+                    key=SearchKey("transaction"), value=token.value, operator=token.operator
+                )
+            )
+        else:
+            # nothing to change append the token as is
+            new_tokens.append(token)
+
+    return new_tokens
+
+
+def _remove_unnecessary_search_filters(tokens: Sequence[QueryToken]) -> Sequence[QueryToken]:
+    """
+    Removes unnecessary tokens that contain `event.type:x` filters.
+    """
+    new_tokens: List[QueryToken] = []
+    for token in tokens:
+        if isinstance(token, SearchFilter):
+            if token.key.name not in ["event.type"]:
+                new_tokens.append(token)
+        elif isinstance(token, ParenExpression):
+            new_tokens.append(ParenExpression(_remove_unnecessary_search_filters(token.children)))
+        else:
+            new_tokens.append(token)
+
+    return new_tokens
+
+
+def get_rule_condition(query: Optional[str]) -> RuleCondition:
+    """
+    Gets the rule condition given a query.
+
+    The rule returned, is in the format which is understood by Relay.
+    """
     try:
         if not query:
             raise UnsupportedSearchQuery(UnsupportedSearchQueryReason.NOT_TRANSACTION_QUERY)
-        else:
-            tokens = parse_search_query(query)
-
-            reason = _is_not_supported(tokens)
-            if reason is not None:
-                raise UnsupportedSearchQuery(reason)
-            # transform a simple message query into a transaction condition:
-            # "foo environment:development" -> "transaction:foo environment:development"
-            tokens = message_to_transaction_condition(tokens)
-            converter = SearchQueryConverter(tokens)
-            condition = converter.convert()
+
+        # First we parse the query.
```

Comment:
I like that you laid out the phases, but I'd probably not include the ordinals in the comments since they can drift pretty easily